xdpderp/xdp.c

```diff
@@ -0,0 +1,513 @@
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			xor6 = (void *)(sa + 1);
+			if ((void *)(xor6 + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+		}
+		xor6->unused = 0x00; // unused byte
+		xor6->family = 0x02;
+		xor6->port = bpf_htons(bpf_ntohs(udp->source) ^ STUN_MAGIC_FOR_PORT_XOR);
+		xor6->addr[0] = bpf_htonl(bpf_ntohl(ip6->saddr.in6_u.u6_addr32[0]) ^ STUN_MAGIC);
+		for (int i = 1; i < 4; i++) {
+			// All three are __be32, no endianness flips.
+			xor6->addr[i] = ip6->saddr.in6_u.u6_addr32[i] ^ req->txid[i-1];
+		}
+	} else {
+		xor = attr_data;
+		if ((void *)(xor + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		xor->unused = 0x00; // unused byte
+		xor->family = 0x01;
+		xor->port = bpf_htons(bpf_ntohs(udp->source) ^ STUN_MAGIC_FOR_PORT_XOR);
+		xor->addr = bpf_htonl(bpf_ntohl(ip->saddr) ^ STUN_MAGIC);
+	}
+
+	// Flip ethernet header source and destination address.
+	__u8 eth_tmp[ETH_ALEN];
+	__builtin_memcpy(eth_tmp, eth->h_source, ETH_ALEN);
+	__builtin_memcpy(eth->h_source, eth->h_dest, ETH_ALEN);
+	__builtin_memcpy(eth->h_dest, eth_tmp, ETH_ALEN);
+
+	// Flip ip header source and destination address.
+	if (is_ipv6) {
+		struct in6_addr ip_tmp = ip6->saddr;
+		ip6->saddr = ip6->daddr;
+		ip6->daddr = ip_tmp;
+	} else {
+		__be32 ip_tmp = ip->saddr;
+		ip->saddr = ip->daddr;
+		ip->daddr = ip_tmp;
+	}
+
+	// Flip udp header source and destination ports;
+	__be16 port_tmp = udp->source;
+	udp->source = udp->dest;
+	udp->dest = port_tmp;
+
+	// Trim packet to end of xor stun attribute.
+	if (is_ipv6) {
+		int shrink = (void *)(xor6 + 1) - data_end;
+		if (bpf_xdp_adjust_tail(ctx, shrink)) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	} else {
+		int shrink = (void *)(xor + 1) - data_end;
+		if (bpf_xdp_adjust_tail(ctx, shrink)) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	}
+
+	// Reset pointers post tail adjustment.
+	data_end = (void *)(long)ctx->data_end;
+	data = (void *)(long)ctx->data;
+	eth = data;
+	if ((void *)(eth + 1) > data_end) {
+		return end(ctx, XDP_ABORTED, is_ipv6);
+	}
+	if (is_ipv6) {
+		ip6 = (void *)(eth + 1);
+		if ((void *)(ip6 + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	} else {
+		ip = (void *)(eth + 1);
+		if ((void *)(ip + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+	}
+
+	// Update ip header total length field and checksum.
+	__u32 cs = 0;
+	if (is_ipv6) {
+		if ((void *)(ip6 +1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		__u16 payload_len = data_end - (void *)(ip6 + 1);
+		ip6->payload_len = bpf_htons(payload_len);
+	} else {
+		__u16 tot_len = data_end - (void *)ip;
+		ip->tot_len = bpf_htons(tot_len);
+		ip->check = 0;
+		cs = bpf_csum_diff(0, 0, (void *)ip, sizeof(*ip), cs);
+		ip->check = csum_fold_flip(cs);
+	}
+
+	// Avoid dynamic length math against the packet pointer, which is just a big
+	// verifier headache. Instead sizeof() all the things.
```

Comment:
I guess this works but it could be tracking the checksum changes all along as the data is updated.

Reply:
It could, but keeping it simpler for now.